File: ios/Runner/TuyaLogic/TuyaHomeHandler.swift
//
//  TuyaHandler.swift
//  Runner
//
//  Created by Jacob Welin - Wejeo on 05/09/2022.
//

import Foundation
import Flutter
import TuyaSmartBaseKit
import TuyaSmartActivatorKit
///
///Access Tuya platform functions
///
class TuyaHomeHandler : NSObject{
    
    static let sharedInstance = TuyaHomeHandler()
    
    let homeManager: TuyaSmartHomeManager
    let tuyaActivator: TuyaSmartActivator
    var currentHome: TuyaSmartHome?
    var flutterResult: FlutterResult?
    var smartDevice :TuyaSmartDevice?
    public var eventSink: FlutterEventSink?
    
    override
    private init(){
        self.homeManager = TuyaSmartHomeManager()
        self.tuyaActivator = TuyaSmartActivator.sharedInstance()
        super.init()
        if LocalDataHandler.currentHomeId != nil {
            self.currentHome = TuyaSmartHome(homeId: LocalDataHandler.currentHomeId!)
            self.initHome(homeId: LocalDataHandler.currentHomeId!)
        }
        
    }
    
    ///
    ///Create a home
    ///
    public func addHome(result:@escaping FlutterResult, homeName: String, geoName: String, roomName: String, lat: Double, lon: Double) {
        
        self.homeManager.addHome(withName: homeName,
                                 geoName: geoName,
                                 rooms: [roomName],
                                 latitude: lat,
                                 longitude: lon,
                                 success: { (homeId) in
            LocalDataHandler.currentHomeId = homeId
            self.initHome(homeId: homeId)
            result(homeId)
        }) { (error) in
            if let e = error {
                result(FlutterError.init(code: " tuyaFailureError", message: "Add home failed: \(e)", details: nil))
            }
        }
    }
    public func removeHome(result:@escaping FlutterResult, homeId: Int64) {
        guard let homeWithID = TuyaSmartHome(homeId: homeId) else{
            result(nil)
            return
        }
        homeWithID.dismiss(success: {
            let message = "Success"
            result(message)
        }) { (error) in
            if let e = error {
                result(FlutterError.init(code: " tuyaFailureError", message: "Remove home failed: \(e)", details: nil))
            }
        }
    }
    
    ///
    ///Get list of homes
    ///
    public func getHomeList(result:@escaping FlutterResult) {
        
        
        self.homeManager.getHomeList(success: { (homes) in
            
            guard let myHomes = homes else{
                result(nil)
                print("nil Homes")
                return
            }
            
            if LocalDataHandler.currentHomeId == nil {
                guard let firstID = myHomes.first?.homeId else{
                    result(nil)
                    return
                }
                print("firstID:")
                print(firstID)
                LocalDataHandler.currentHomeId = firstID
                self.currentHome = TuyaSmartHome(homeId: firstID)
                self.initHome(homeId: firstID)
            }
            
            let sortedHomes = myHomes.sorted { $1.homeId == self.currentHome?.homeModel.homeId ? false : $0.homeId == self.currentHome?.homeModel.homeId ? true : $0.homeId < $1.homeId }
            
            let homeDictList = sortedHomes.reduce(into: [[AnyHashable: Any]]()) { array, value in
                array.append(["name": value.name!, "homeId" : value.homeId, "geoName" : value.geoName!, "lat" : value.latitude, "lon" : value.longitude])
            }
            
            result(homeDictList.toJSONString())
        }) { (error) in
            if let e = error {
                result(FlutterError.init(code: " tuyaFailureError", message: "Get home list failed: \(e)", details: nil))
            }
        }
    }
    
    
    public func setCurrentHome(result:@escaping FlutterResult, homeId:Int64) {
        LocalDataHandler.currentHomeId = homeId;
        self.initHome(homeId: homeId)
        result(nil)
    }
    
    public func getCurrentHome(result:@escaping FlutterResult) {
        
        print(LocalDataHandler.currentHomeId ?? "null home")
        guard let homeID = LocalDataHandler.currentHomeId else{
            print("currentHome nil")
            result(0)
            return
        }
        result(homeID)
    }
    
}

File: android/app/src/main/kotlin/dk/wejeo/wejeo_smart/tuya_logic/TuyaHomeHandler.kt
package dk.wejeo.wejeo_smart.tuya_logic

import android.os.Handler
import android.os.Looper
import android.util.Log
import com.tuya.smart.home.sdk.TuyaHomeSdk
import com.tuya.smart.home.sdk.api.ITuyaHome
import com.tuya.smart.home.sdk.bean.HomeBean
import com.tuya.smart.home.sdk.callback.ITuyaGetHomeListCallback
import com.tuya.smart.home.sdk.callback.ITuyaHomeResultCallback
import com.tuya.smart.sdk.api.IResultCallback
import dk.wejeo.wejeo_smart.LocalDataHandler
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodChannel
import org.json.JSONArray
import org.json.JSONObject
import java.util.*
import kotlin.concurrent.schedule


open class TuyaHomeHandler {
    val LOG_TAG = "HomeConfig_#JW"

    companion object {
        var eventSink: EventChannel.EventSink? = null
        private val uiThreadHandler: Handler = Handler(Looper.getMainLooper())

        //        var currentHome: HomeBean? = null
        var mHome: ITuyaHome? = null

        fun updateHomeData() {
            if (mHome?.homeBean == null) {
                return
            } else {
                val dataString = homeDataToJson(mHome?.homeBean)
                uiThreadHandler.post {
                    eventSink?.success(dataString)
                }
            }

        }

        fun homeDataToJson(bean: HomeBean?): String {
            val deviceList = bean?.deviceList
            val devicesListMap = mutableListOf<Map<String, Any?>>()
            if (deviceList != null) {
                for (device in deviceList) {

                    val data: Map<String, Any?> =
                        mapOf(
                            "name" to device.name,
                            "devId" to device.devId,
                            "isOnline" to device.isOnline,
                            "isCloudOnline" to device.isCloudOnline,
                            "onlineType" to 0,
                            "deviceType" to 0,
                            "dps" to JSONObject(device.dps).toString(),
                            "homeId" to 0,
                            "roomId" to 0,
                        )
                    devicesListMap.add(data)
                }
                val jsArray = JSONArray(devicesListMap)
                return jsArray.toString()
            } else {
                return "[]"
            }
        }
    }
//    static let sharedInstance = TuyaHomeHandler()

//    val homeManager: TuyaSmartHomeManager
//    val tuyaActivator: TuyaSmartActivator
//    var currentHome: TuyaSmartHome?
//    var flutterResult: FlutterResult?
//    var smartDevice :TuyaSmartDevice?

//    override
//    private init(){
//        self.homeManager = TuyaSmartHomeManager()
//        self.tuyaActivator = TuyaSmartActivator.sharedInstance()
//        super.init()
//        if LocalDataHandler.currentHomeId != nil {
//            self.currentHome = TuyaSmartHome(homeId: LocalDataHandler.currentHomeId!)
//            self.initHome(homeId: LocalDataHandler.currentHomeId!)
//        }
//
//    }

    ///
    ///Create a home
    ///
    fun addHome(
        result: MethodChannel.Result,
        homeName: String,
        geoName: String,
        roomName: String,
        lat: Double,
        lon: Double
    ) {
        TuyaHomeSdk.getHomeManagerInstance()
            .createHome(
                homeName,
                lon,
                lat,
                geoName,
                listOf(roomName),
                object : ITuyaHomeResultCallback {
                    override fun onError(code: String?, error: String?) {
                        result.error(code.toString(), error, "Failed to add home")
                    }

                    override fun onSuccess(bean: HomeBean) {
                        LocalDataHandler.currentHomeId = bean.homeId
                        //TODO self.initHome(homeId: homeId)
                        result.success(bean.homeId)
                    }
                })
    }

    fun removeHome(result: MethodChannel.Result, homeId: Long) {

        TuyaHomeSdk.newHomeInstance(homeId).dismissHome(object : IResultCallback {
            override fun onSuccess() {
                // do something
                result.success("Success")
            }

            override fun onError(code: String?, error: String?) {
                result.error(code.toString(), error, "Failed to remove home")
            }
        })
    }


    ///
    ///Get list of homes
    ///
    fun getHomeList(result: MethodChannel.Result) {
        TuyaHomeSdk.getHomeManagerInstance().queryHomeList(object : ITuyaGetHomeListCallback {
            override fun onSuccess(homeBeans: List<HomeBean>) {
                // do something
                if (homeBeans.isEmpty()) {
                    result.success(null)
                    return
                }
                var firstID = homeBeans.first().homeId
                LocalDataHandler.currentHomeId = firstID

                //TODO
//                self.currentHome = TuyaSmartHome(homeId: firstID)
//                self.initHome(homeId: firstID)

                var homeDictList: List<Map<String, Any?>> = homeBeans.map {
                    mapOf(
                        "name" to it.name,
                        "homeId" to it.homeId,
                        "geoName" to it.geoName,
                        "lat" to it.lat,
                        "lon" to it.lon
                    )
                }
                val jsArray = JSONArray(homeDictList)
                Log.i(LOG_TAG, "homedata: $jsArray")
                result.success(jsArray.toString())
            }

            override fun onError(code: String?, error: String?) {
                result.error(code.toString(), error, "Failed to get home list")
            }
        })

    }


    fun setCurrentHome(result: MethodChannel.Result, homeId: Long) {
        LocalDataHandler.currentHomeId = homeId;
        //TODO
//        self.initHome(homeId: homeId)
        result.success(null)
    }

    fun getCurrentHome(result: MethodChannel.Result) {
        val homeId = LocalDataHandler.currentHomeId
        if (homeId == null) {
            result.success(0)
            return
        }
        result.success(homeId)
    }

    fun updateHomeData(result: MethodChannel.Result) {
        updateHomeData()
        result.success(null)
    }
}